crates/wasi-http/src/types_impl.rs

```diff
@@ -165,35 +160,33 @@ impl crate::types::Host for WasiHttp {
         authority: String,
         headers: Headers,
     ) -> wasmtime::Result<OutgoingRequest> {
-        let id = self.request_id_base;
-        self.request_id_base = self.request_id_base + 1;
-
-        let mut req = ActiveRequest::new(id);
+        let mut req = ActiveRequest::new();
         req.path = path;
         req.query = query;
         req.authority = authority;
         req.method = method;
-        req.headers = match self.fields.get(&headers) {
-            Some(h) => h.clone(),
-            None => bail!("headers not found."),
-        };
+        req.headers = self.table().get_fields(headers).map_err(convert)?.clone();
         req.scheme = scheme;
-        self.requests.insert(id, req);
+        let id = self.push_request(Box::new(req)).map_err(convert)?;
         Ok(id)
     }
     fn outgoing_request_write(
         &mut self,
         request: OutgoingRequest,
     ) -> wasmtime::Result<Result<OutgoingStream, ()>> {
-        let mut req = self
-            .requests
-            .get_mut(&request)
-            .ok_or_else(|| anyhow!("unknown request: {request}"))?;
-        if req.body == 0 {
-            req.body = self.streams_id_base;
-            self.streams_id_base = self.streams_id_base + 1;
-        }
-        Ok(Ok(req.body))
+        let req = self.table().get_request(request).map_err(convert)?;
+        Ok(Ok(req.body().unwrap_or_else(|| {
+            let buf = ByteStream::new();
+            let new = self
+                .push_output_stream(Box::new(buf))
+                .expect("valid output stream");
+            let req = self
+                .table_mut()
+                .get_request_mut(request)
+                .expect("request to be found");
+            req.set_body(new);
+            new
+        })))
```

Comment:
How about extracting this to a `let body = req.body().unwrap_or_else(/* ... */)` for clarity?